fe/src/main/java/org/apache/doris/clone/ColocateTableBalancer.java

```diff
@@ -507,111 +508,133 @@ private Long selectCloneBackendIdForRemove(com.google.common.collect.Table<Long,
-            for (long tableId : allTableIds) {
-                OlapTable olapTable = (OlapTable) db.getTable(tableId);
-                for (Partition partition : olapTable.getPartitions()) {
-                    replicateNum = olapTable.getPartitionInfo().getReplicationNum(partition.getId());
-                    for (MaterializedIndex index : partition.getMaterializedIndices()) {
-                        List<Tablet> tablets = index.getTablets();
-                        for (int i = 0; i < tablets.size() && i < needMoveBucketSeqs; i++) {
-                            Tablet tablet = tablets.get(i);
-                            List<Replica> replicas = tablet.getReplicas();
-                            List<Long> sortedReplicaIds = sortReplicaId(replicas);
-                            //always delete replica which id is minimum
-                            Replica deleteReplica = tablet.getReplicaById(sortedReplicaIds.get(0));
+            List<List<Long>> newBackendsPerBucketSeq = balance(backendsPerBucketSeq, addedBackendIds);
+            markGroupBalancing(groupId);
+            persistBackendsToBucketSeqMeta(groupId, newBackendsPerBucketSeq);
+        } catch (Exception e) {
+            LOG.error(e.getMessage(), e);
+        } finally {
+            db.readUnlock();
+        }
+    }
 
-                            long tabletSizeB = deleteReplica.getDataSize() * partition.getMaterializedIndices().size()
-                             * olapTable.getPartitions().size() * allTableIds.size();
-                            CloneTabletInfo tabletInfo = new CloneTabletInfo(db.getId(), tableId, partition.getId(),
-                            index.getId(), tablet.getId(), (short) replicateNum, (short) replicateNum,
-                            tabletSizeB, tablet.getBackendIds());
+    /**
+     * Returns a map that the key is backend id, the value is backend counter
+     * The map will sort by counter in descending order
+     *
+     * @param backends the backend id list
+     * @return a descending sorted map
+     */
+    private static Map<Long, Long> getBackendCounter(List<Long> backends) {
+        Map<Long, Long> backendCounter = backends.stream()
+                .collect(Collectors.groupingBy(Function.identity(), Collectors.counting()));
 
-                            Long cloneReplicaBackendId = newGroup2BackendsPerBucketSeq.get(groupId, i);
-                            if (cloneReplicaBackendId == null) {
-                                // select dest backend
-                                cloneReplicaBackendId = addedBackendIds.get(i % addedBackendIds.size());
-                                newGroup2BackendsPerBucketSeq.put(groupId, i, cloneReplicaBackendId);
-                            }
+        return backendCounter.entrySet().stream().sorted(Collections.reverseOrder(Map.Entry.comparingByValue()))
+                .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue, (e1, e2) -> e1, LinkedHashMap::new));
```

Comment:
Hash map is not ordered, I think you what you need is List<Map.Entry<Long, Long>>

Reply:
The input order and out order for `LinkedHashMap` is the same.

Reply:
OK, I see